### PR TITLE
Fix broadcast_all crashing on Tensor-likes

### LINKED ISSUE #37141: torch.distributions.utils.broadcast_all does not support Tensor-like objects
## 🐛 Bug

`torch.distributions.utils.broadcast_all` will crash when it receives Tensor-like objects, as `torch.is_tensor` returns `False` on Tensor-likes that do not subclass Tensor. This prevents creating distribution objects using Tensor-likes.

## To Reproduce

Consider the class

```python
class TensorLike:
    def __torch_function__(func, types, args=(), kwargs=None):
        print("In __torch_function_")
```
Now create (for example) a Normal distribution using it:

```python
from torch.distributions import Normal

Normal(loc=TensorLike(), scale=1)
```

This crashes with
```
Traceback (most recent call last):
  File "/Users/emile/Documents/GitHub/storchastic/test/test_broadcast_all.py", line 11, in <module>
    Normal(TensorLike(), scale=1)
  File "/Users/emile/.conda/envs/threes-ai/lib/python3.7/site-packages/torch/distributions/normal.py", line 44, in __init__
    self.loc, self.scale = broadcast_all(loc, scale)
  File "/Users/emile/.conda/envs/threes-ai/lib/python3.7/site-packages/torch/distributions/utils.py", line 24, in broadcast_all
    raise ValueError('Input arguments must all be instances of numbers.Number or torch.tensor.')
ValueError: Input arguments must all be instances of numbers.Number or torch.tensor.
```

## Expected behavior

If a Tensor-like is given to `broadcast_all`, the `torch.broadcast_tensors` call will defer handling of Tensor-likes to `__torch_function__`. Tensor-likes should pass the first check in `broadcast_all`. Tensor-likes should be ignored when creating `Tensor` objects from `Number`s. This should allow Tensor-likes to be used in `broadcast_all`.

## Environment

PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.4
GCC version: Could not collect
CMake version: version 3.14.5

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] storchastic==0.0.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      233  
[conda] mkl-service               2.3.0            py37hfbe908c_0  
[conda] mkl_fft                   1.0.14           py37h5e564d8_0  
[conda] mkl_random                1.1.0            py37ha771720_0  
[conda] numpy                     1.17.2           py37h99e6662_0  
[conda] numpy-base                1.17.2           py37h6575580_0  
[conda] numpydoc                  0.9.1                      py_0  
[conda] torch                     1.3.1                    pypi_0    pypi

## Additional context

`torch.is_tensor` is in the list of `get_ignored_functions`, meaning that it will not call `__torch_function__` by design and thus will not allow Tensor-likes to pass by design.

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

### PR DESCRIPTION
**Pull request by HEmile**
This ensures Tensor-likes that implement `__torch_function__` are properly handled by `torch.distributions.utils.broadcast_all`.  See Issue #37141 .

In this implementation, Number's will not be cast to the dtype of Tensor-likes.

Fixes #37141

### CODE REVIEW

**Comment by HEmile** [comment]
I just signed the CLA, not sure how to recheck this.

**Review by walterddr — COMMENTED** [review]
thanks for the contribution @HEmile. could you add a test (e.g. test_distributions.py or test_autograd.py) to validate your change for tensor-like broadcast?

**Comment by HEmile** [comment]
I chose to implement the test in `test_overrides` as it already implements a `Wrapper` type that could be used to test the behavior. Let me know if this is not the right place.

File: test/test_overrides.py [hunk 1]
```diff
@@ -838,5 +838,27 @@ def test_newones(self):
         n = t.new_ones((1, 2))
         self.assertEqual(type(n), SubTensor2)
 
+class TestBroadcastAllOverride(TestCase):
+    " Regression test for gh-48169"
```
Comment (walterddr) [inline test/test_overrides.py:842]: let's use the actual issue link instead of the PR link
```suggestion
    " test for gh-37141
```

**Review by walterddr — APPROVED** [review]
lgtm . thank you so much for your changes

File: torch/distributions/utils.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import torch
 import torch.nn.functional as F
 from typing import Dict, Any
-
+from torch.overrides import has_torch_function
 
 def broadcast_all(*values):
```
Comment (malfet) [inline torch/distributions/utils.py:8]: Would it be possible to add type annotation here (not sure what is the type of torch.Function, I guess `Callable[torch.Tensor]`)?
```suggestion
def broadcast_all(*values: Union[torch.Tensor, Number]):
```

**Review by malfet — APPROVED** [review]

File: torch/distributions/utils.py (hunk 2 above)
Comment (HEmile) [reply torch/distributions/utils.py:8]: I don't think there is a type annotation for objects that implement __torch_function__. Unless PyTorch has implemented an abstract class for this somewhere, but in this case it's not well documented.
And this suggestion would give the impression that Tensor-Likes wouldn't be allowed.

File: test/test_overrides.py [hunk 3]
```diff
@@ -838,5 +838,27 @@ def test_newones(self):
         n = t.new_ones((1, 2))
         self.assertEqual(type(n), SubTensor2)
 
+class TestBroadcastAllOverride(TestCase):
+    " test for gh-37141
```
Comment (walterddr) [inline test/test_overrides.py:842]:
```suggestion
    """test fo rgh-37141"""
```

Comment (HEmile) [reply test/test_overrides.py]: This introduced a typo, so I created a different commit

**Comment by walterddr** [comment]
thanks for the change @HEmile and sorry for the late response during the holiday. do you mind rebasing the PR to latest viable/strict so that those failed CI tests could be rerun correctly? Thanks

**Pull request closed without merging**